File: ruby/triangle/triangle.rb
class Triangle
  def initialize(sides)
    @a = sides[0]
    @b = sides[1]
    @c = sides[2]
  end

  def equilateral?
    triangle? && a == b && b == c
  end

  def isosceles?
    triangle? && (a == b || b == c || c == a)
  end

  def scalene?
    triangle? && (a != b && b != c)
  end

  private

  attr_reader :a, :b, :c

  def triangle?
    [a, b, c].all?(&:positive?) && combinations.all? { |(a, b, c)| a + b >= c }
  end

  def combinations
    [[a, b, c], [b, c, a], [c, a, b]]
  end
end

File: haskell/triangle/src/Triangle.hs
module Triangle
  ( TriangleType(..)
  , triangleType
  )
where

data TriangleType = Equilateral
                  | Isosceles
                  | Scalene
                  | Illegal
                  deriving (Eq, Show)

triangleType :: (Num a, Ord a) => a -> a -> a -> TriangleType
triangleType a b c | not $ isLegal a b c        = Illegal
                   | a == b && b == c           = Equilateral
                   | a == b || b == c || a == c = Isosceles
                   | otherwise                  = Scalene

isLegal :: (Num a, Ord a) => a -> a -> a -> Bool
isLegal a b c = hasSides && hasTriangleInequality
 where
  hasSides              = a > 0 && b > 0 && c > 0
  hasTriangleInequality = a + b > c && a + c > b && b + c > a
